File: Code/Projects/Spec_driven_TextFinder/Rust_Spec_driven_TextFinder/Rust_Spec_driven_Cmdline/src/unit_tests.rs
//! Unit suite for rust_textfinder_cmdline, exercising Spec_Rust_TextFinder_Cmdline.md.

use super::*;

fn args(items: &[&str]) -> Vec<String> {
    let mut all = vec![String::from("rust_textfinder")];
    all.extend(items.iter().map(|s| String::from(*s)));
    all
}

fn parsed(items: &[&str]) -> ProgramCommands {
    match parse(&args(items)) {
        Ok(commands) => commands,
        Err(diag) => panic!("expected success, got diagnostic:\n{diag}"),
    }
}

fn reason(items: &[&str]) -> String {
    match parse(&args(items)) {
        Ok(_) => panic!("expected a diagnostic, parse succeeded"),
        Err(diag) => diag.lines().next().unwrap_or("").to_string(),
    }
}

// --- defaults, section 4 ---

#[test]
fn empty_command_line_yields_the_defaults() {
    let commands = parsed(&[]);
    let defaults = ProgramCommands::default();
    assert_eq!(commands.root_paths, defaults.root_paths);
    assert_eq!(commands.root_paths, vec![String::from(".")]);
    assert!(commands.extensions.is_empty());
    assert_eq!(commands.regex_text, ".");
    assert!(commands.recurse);
    assert!(commands.suppress_on_no_match);
    assert!(!commands.verbose);
    assert!(!commands.help);
    assert!(!commands.line_numbers);
    assert!(!commands.matched_line);
}

// --- switch syntax, section 5 rules 1-3 ---

#[test]
fn both_introducers_are_equivalent() {
    assert!(!parsed(&["/s", "false"]).recurse);
    assert!(!parsed(&["-s", "false"]).recurse);
}

#[test]
fn booleans_are_matched_case_insensitively() {
    assert!(parsed(&["/n", "TRUE"]).line_numbers);
    assert!(parsed(&["/n", "True"]).line_numbers);
    assert!(!parsed(&["/h", "FALSE"]).suppress_on_no_match);
}

#[test]
fn switch_letters_are_case_sensitive() {
    assert!(parsed(&["/H", "true"]).help);
    assert!(parsed(&["/H", "true"]).suppress_on_no_match);
    assert!(!parsed(&["/h", "false"]).help);
    assert!(!parsed(&["/h", "false"]).suppress_on_no_match);
}

#[test]
fn an_argument_is_consumed_verbatim_even_when_it_looks_like_a_switch() {
    let commands = parsed(&["/r", "-P"]);
    assert_eq!(commands.regex_text, "-P");
}

#[test]
fn program_name_is_not_inspected() {
    let mut given = vec![String::from("/s")];
    given.push(String::from("/n"));
    given.push(String::from("true"));
    let commands = parse(&given).expect("args[0] must be skipped");
    assert!(commands.line_numbers);
}

// --- accumulation, section 5 rule 4 ---

#[test]
fn repeated_root_paths_accumulate_in_order() {
    let commands = parsed(&["/P", "src", "/P", "docs", "/P", "."]);
    assert_eq!(commands.root_paths, vec!["src", "docs", "."]);
}

#[test]
fn first_root_path_replaces_the_default_even_when_equal_to_it() {
    let commands = parsed(&["/P", ".", "/P", "src"]);
    assert_eq!(commands.root_paths, vec![".", "src"]);
}

#[test]
fn last_occurrence_wins_for_every_other_switch() {
    let commands = parsed(&["/r", "first", "/r", "second", "/s", "true", "/s", "false"]);
    assert_eq!(commands.regex_text, "second");
    assert!(!commands.recurse);
}

// --- diagnostics, section 6 ---

#[test]
fn a_token_without_an_introducer_is_not_a_switch() {
    assert_eq!(reason(&["P", "src"]), "not a switch: P");
    assert_eq!(reason(&[""]), "not a switch: ");
}

#[test]
fn an_undefined_introducer_led_token_is_unrecognized() {
    assert_eq!(reason(&["/x", "1"]), "unrecognized switch: /x");
    assert_eq!(reason(&["/"]), "unrecognized switch: /");
    assert_eq!(reason(&["-"]), "unrecognized switch: -");
    assert_eq!(reason(&["/PP", "src"]), "unrecognized switch: /PP");
}

#[test]
fn a_trailing_switch_is_missing_its_argument() {
    assert_eq!(reason(&["/P"]), "missing argument for switch: /P");
    assert_eq!(reason(&["/s", "true", "-n"]), "missing argument for switch: -n");
}

#[test]
fn a_boolean_switch_rejects_any_other_value() {
    assert_eq!(reason(&["/s", "yes"]), "invalid boolean for /s: yes");
    assert_eq!(reason(&["-L", "1"]), "invalid boolean for -L: 1");
}

#[test]
fn empty_arguments_are_rejected_for_root_path_and_expression() {
    assert_eq!(reason(&["/P", ""]), "empty root path for switch: /P");
    assert_eq!(reason(&["-r", ""]), "empty expression for switch: -r");
}

#[test]
fn every_diagnostic_ends_with_the_usage_line() {
    let diag = parse(&args(&["/x", "1"])).expect_err("must fail");
    assert_eq!(diag, format!("unrecognized switch: /x\n{}", usage_line()));
    assert!(diag.ends_with('\n'));
}

#[test]
fn parsing_stops_at_the_first_violation() {
    assert_eq!(reason(&["/s", "maybe", "/x", "1"]), "invalid boolean for /s: maybe");
}

// --- extension normalization, section 7 ---

#[test]
fn extensions_are_split_trimmed_and_stripped_of_one_dot() {
    let commands = parsed(&["/p", " cpp, .rs ,h"]);
    assert_eq!(commands.extensions, vec!["cpp", "rs", "h"]);
}

#[test]
fn empty_extension_items_are_discarded() {
    assert_eq!(parsed(&["/p", "cpp,,rs"]).extensions, parsed(&["/p", "cpp, rs"]).extensions);
    assert!(parsed(&["/p", ""]).extensions.is_empty());
    assert!(parsed(&["/p", " , . , "]).extensions.is_empty());
}

#[test]
fn only_one_leading_dot_is_stripped() {
    assert_eq!(parsed(&["/p", "..cpp"]).extensions, vec![".cpp"]);
}

#[test]
fn the_six_named_characters_are_trimmed_and_no_others() {
    let commands = parsed(&["/p", " \t\n\u{000B}\u{000C}\rcpp \t\r, \u{00A0}rs"]);
    assert_eq!(commands.extensions, vec!["cpp", "\u{00A0}rs"]);
}

#[test]
fn duplicate_extensions_are_retained_and_order_preserved() {
    assert_eq!(parsed(&["/p", "rs, cpp, rs"]).extensions, vec!["rs", "cpp", "rs"]);
}

#[test]
fn extension_case_is_not_folded_by_the_parser() {
    assert_eq!(parsed(&["/p", "CPP"]).extensions, vec!["CPP"]);
}

// --- help text and option listing, section 8 ---

#[test]
fn usage_line_is_the_first_line_of_the_help_text() {
    let help = help_text();
    assert!(help.starts_with(&usage_line()));
    assert_eq!(help.lines().next(), usage_line().lines().next());
    assert!(usage_line().starts_with("usage: rust_textfinder [/P path]"));
    assert!(usage_line().ends_with("[/n bool] [/L bool]\n"));
}

#[test]
fn help_text_carries_every_switch_and_ends_with_a_newline() {
    let help = help_text();
    for switch in ["/P", "/p", "/r", "/s", "/h", "/v", "/H", "/n", "/L"] {
        assert!(help.contains(&format!("\n  {switch}  ")), "help omits {switch}");
    }
    assert!(help.ends_with("searching.\n"));
}

#[test]
fn option_listing_of_a_bare_v_true_command_line() {
    let commands = parsed(&["-v", "true"]);
    assert_eq!(
        options_text(&commands),
        "/P .\n/p\n/r .\n/s true\n/h true\n/v true\n/H false\n/n false\n/L false\n"
    );
}

#[test]
fn option_listing_of_the_defaults_reads_v_false() {
    assert_eq!(
        options_text(&ProgramCommands::default()),
        "/P .\n/p\n/r .\n/s true\n/h true\n/v false\n/H false\n/n false\n/L false\n"
    );
}

#[test]
fn option_listing_emits_one_line_per_root_and_a_joined_extension_list() {
    let commands = parsed(&["/P", "src", "/P", "docs", "/p", ".rs, cpp", "/r", "int\\s+main"]);
    let text = options_text(&commands);
    assert!(text.starts_with("/P src\n/P docs\n/p rs, cpp\n/r int\\s+main\n"));
}

#[test]
fn no_line_of_the_option_listing_ends_in_whitespace() {
    let commands = parsed(&["/p", ""]);
    for line in options_text(&commands).lines() {
        assert!(!line.ends_with(' '), "trailing space on: {line:?}");
    }
    assert!(options_text(&commands).contains("\n/p\n"));
}

#[test]
fn every_returned_string_ends_with_a_newline() {
    assert!(usage_line().ends_with('\n'));
    assert!(help_text().ends_with('\n'));
    assert!(options_text(&ProgramCommands::default()).ends_with('\n'));
}

File: Code/Projects/Spec_driven_TextFinder/CSharp_Spec_driven_TextFinder/CSharp_Spec_driven_Cmdline/test/UnitTests.cs
// UnitTests.cs - unit suite for CSharp_TextFinder_Cmdline, exercising Spec_CSharp_TextFinder_Cmdline.md
//
// Spec_TextFinder.md §6 admits no third-party package, so the harness is the Check class
// below. It is duplicated in each component's suite rather than shared, so that a
// component can be read, built, and tested alone.

using CSharp_TextFinder_Cmdline;

namespace CSharp_TextFinder_Cmdline_UnitTest;

internal sealed class Check
{
    private int _failures;
    private int _total;

    public void That(bool ok, string name) => Record(ok, name);

    public void Equal(string actual, string expected, string name)
    {
        bool ok = actual == expected;
        Record(ok, name);
        if (ok) return;
        Console.WriteLine("          expected: [" + Visible(expected) + "]");
        Console.WriteLine("          actual:   [" + Visible(actual) + "]");
    }

    private void Record(bool ok, string name)
    {
        ++_total;
        if (!ok) ++_failures;
        Console.WriteLine((ok ? "  PASS  " : "  FAIL  ") + name);
    }

    public int Report()
    {
        Console.WriteLine($"  {_total - _failures} of {_total} passed");
        Console.WriteLine(_failures == 0 ? "PASS" : "FAIL");
        return _failures;
    }

    // The CRLF failures of a first run look identical to their expected text without this.
    private static string Visible(string text) =>
        text.Replace("\r", "\\r").Replace("\n", "\\n");
}

internal static class UnitTests
{
    private static ProgramCommands Parsed(params string[] args)
    {
        if (!CommandLine.TryParse(args, out ProgramCommands? commands, out string? diagnostic))
        {
            throw new InvalidOperationException("expected success, got: " + diagnostic);
        }

        return commands;
    }

    private static string Reason(params string[] args)
    {
        if (CommandLine.TryParse(args, out _, out string? diagnostic))
        {
            throw new InvalidOperationException("expected a diagnostic, parse succeeded");
        }

        return diagnostic.Split('\n')[0];
    }

    private static int Main()
    {
        var check = new Check();
        Console.WriteLine("CSharp_TextFinder_Cmdline unit tests");

        // --- defaults, §4 ---
        ProgramCommands defaults = Parsed();
        check.That(defaults.RootPaths.Count == 1 && defaults.RootPaths[0] == ".", "default /P is .");
        check.That(defaults.Extensions.Count == 0, "default /p is empty");
        check.Equal(defaults.RegexText, ".", "default /r is .");
        check.That(defaults.Recurse, "default /s is true");
        check.That(defaults.SuppressOnNoMatch, "default /h is true");
        check.That(!defaults.Verbose, "default /v is false");
        check.That(!defaults.Help, "default /H is false");
        check.That(!defaults.LineNumbers, "default /n is false");
        check.That(!defaults.MatchedLine, "default /L is false");
        check.That(new ProgramCommands().RegexText == defaults.RegexText,
            "a new ProgramCommands equals the result of parsing an empty array");

        // --- §5: the scan begins at index 0 ---
        check.Equal(Parsed("-r", "alpha").RegexText, "alpha", "the first element is a switch, not a program name");
        check.Equal(Reason("CSharp_TextFinder", "-r", "alpha"), "not a switch: CSharp_TextFinder",
            "a program name in args[0] is refused, since Main's array carries none");

        // --- §5 rules 1 through 3 ---
        check.That(Parsed("-s", "false").Recurse == false && Parsed("/s", "false").Recurse == false,
            "introducers / and - are equivalent");
        check.That(Parsed("-s", "FALSE").Recurse == false && Parsed("-s", "True").Recurse,
            "boolean values fold case");
        check.That(Parsed("-h", "false").SuppressOnNoMatch == false && !Parsed("-H", "false").Help,
            "/h and /H are distinct switches");
        check.Equal(Parsed("-r", "one", "-r", "two").RegexText, "two", "last occurrence wins for /r");
        check.Equal(Parsed("-r", "-n").RegexText, "-n",
            "an argument beginning with an introducer is taken verbatim");

        // --- §5 rule 4 ---
        ProgramCommands roots = Parsed("-P", "one", "-P", "two");
        check.That(roots.RootPaths.Count == 2 && roots.RootPaths[0] == "one" && roots.RootPaths[1] == "two",
            "/P accumulates in the order given");
        ProgramCommands dot = Parsed("-P", ".");
        check.That(dot.RootPaths.Count == 1, "the first /P replaces the default even when equal to it");

        // --- §7: normalization ---
        check.Equal(string.Join("|", Parsed("-p", " .cs , , txt ").Extensions), "cs|txt",
            "items are split, trimmed, and stripped of one dot, and empties discarded");
        check.Equal(string.Join("|", Parsed("-p", "..cs").Extensions), ".cs",
            "only one leading dot is stripped");
        check.Equal(string.Join("|", Parsed("-p", ". cs, .\ttxt ").Extensions), "cs|txt",
            "an item is trimmed again after its dot is stripped");
        check.That(CommandLine.OptionsText(Parsed("-p", ". cs")).Contains("\n/p cs\n"),
            "the second trim keeps stray whitespace out of the §5.3 listing line");
        check.Equal(string.Join("|", Parsed("-p", ".cs, cs").Extensions), "cs|cs",
            "duplicates are retained");
        check.That(Parsed("-p", " , , ").Extensions.Count == 0, "an all-separator list is empty");
        check.Equal(string.Join("|", Parsed("-p", "CS").Extensions), "CS",
            "extension case is not folded by the parser");
        check.Equal(
            string.Join("|", Parsed("-p", "a" + (char)0x0B + ",\tb\n,\rc ,d" + (char)0x0C).Extensions),
            "a|b|c|d",
            "the six characters §5 names are trimmed");
        check.Equal(string.Join("|", Parsed("-p", "\u00a0cs").Extensions), "\u00a0cs",
            "no-break space is not trimmed, since §5 does not name it");

        // --- §6: the six reason lines ---
        check.Equal(Reason("P", "."), "not a switch: P", "not a switch");
        check.Equal(Reason("/x", "1"), "unrecognized switch: /x", "unrecognized switch");
        check.Equal(Reason("/ss", "1"), "unrecognized switch: /ss", "an over-long token is unrecognized");
        check.Equal(Reason("/", "1"), "unrecognized switch: /", "a bare introducer is unrecognized");
        check.Equal(Reason("/P"), "missing argument for switch: /P", "missing argument");
        check.Equal(Reason("-s", "yes"), "invalid boolean for -s: yes", "invalid boolean");
        check.Equal(Reason("/P", ""), "empty root path for switch: /P", "empty root path");
        check.Equal(Reason("/r", ""), "empty expression for switch: /r", "empty expression");
        check.Equal(Reason("-s", "yes", "/x", "1"), "invalid boolean for -s: yes",
            "parsing stops at the first violation");

        CommandLine.TryParse(new[] { "/P" }, out _, out string? diagnostic);
        check.That(diagnostic!.EndsWith(CommandLine.UsageLine(), StringComparison.Ordinal),
            "every diagnostic ends with the usage line");
        check.That(!diagnostic.Contains('\r'), "a diagnostic carries LF and no CR");

        // --- §8: help text ---
        string help = CommandLine.HelpText();
        check.That(help.StartsWith(CommandLine.UsageLine(), StringComparison.Ordinal),
            "the help text begins with the usage line");
        check.That(help.Contains("usage: CSharp_TextFinder [", StringComparison.Ordinal),
            "the usage line names the executable");
        check.That(help.EndsWith("searching.\n", StringComparison.Ordinal), "the help text ends with a newline");
        check.That(!help.Contains('\r'), "the help text carries LF and no CR");
        check.That(help.Split('\n').Length - 1 == 22, "the help text is the 22 lines §5.1 fixes");
        check.That(help.Contains("counting the files and directories it reached.", StringComparison.Ordinal),
            "help names the run summary of §3.6");
        foreach (string letter in new[] { "/P", "/p", "/r", "/s", "/h", "/v", "/H", "/n", "/L" })
        {
            check.That(help.Contains("  " + letter + "  ", StringComparison.Ordinal), "help lists " + letter);
        }

        check.That(CommandLine.UsageLine().Split('\n').Length - 1 == 1,
            "the usage line is one line");

        // --- §8: option listing ---
        check.Equal(CommandLine.OptionsText(new ProgramCommands()),
            "/P .\n/p\n/r .\n/s true\n/h true\n/v false\n/H false\n/n false\n/L false\n",
            "the listing of the defaults is the nine lines of §5.3, with /v false");
        check.Equal(CommandLine.OptionsText(Parsed("-v", "true")),
            "/P .\n/p\n/r .\n/s true\n/h true\n/v true\n/H false\n/n false\n/L false\n",
            "the listing of a bare -v true is the nine lines §5.3 gives");
        check.Equal(CommandLine.OptionsText(Parsed("-P", "one", "-P", "two", "-p", "cs, md")).Split('\n')[0],
            "/P one", "one /P line per root path, in traversal order");
        check.That(CommandLine.OptionsText(Parsed("-p", "cs, md")).Contains("/p cs, md", StringComparison.Ordinal),
            "the extension list is joined by comma and space");
        check.That(!CommandLine.OptionsText(new ProgramCommands()).Contains("True", StringComparison.Ordinal),
            "booleans render lower case, so bool.ToString is not used");
        string listing = CommandLine.OptionsText(Parsed("-r", "a b"));
        foreach (string line in listing.Split('\n'))
        {
            if (line.Length == 0) continue;
            check.That(line.Trim() == line, "no line of the listing ends in whitespace: [" + line + "]");
        }

        check.That(CommandLine.OptionsText(Parsed("-r", "a\\.b")).Contains("/r a\\.b", StringComparison.Ordinal),
            "/r renders the expression verbatim, backslash included");

        return check.Report();
    }
}
